web/popminer/dispatch.go

```diff
@@ -0,0 +1,438 @@
+
+	// Create and return the Promise object
+	return promiseConstructor.New(handler)
+}
+
+func parseArgs(args []js.Value) (*Dispatch, error) {
+	// Verify we received data readable command
+	if len(args) != 1 {
+		return nil, fmt.Errorf("expected 1 argument, got %v", len(args))
+	}
+	data := args[0]
+	if data.Type() != js.TypeObject {
+		return nil, fmt.Errorf("expected an object, got: %v", data.Type())
+	}
+
+	m := data.Get("method")
+	if m.Type() != js.TypeString {
+		return nil, fmt.Errorf("expected method to be a string, got: %v", m.Type())
+	}
+	d, ok := Dispatcher[m.String()]
+	if !ok {
+		log.Warningf("method not found: %v", m.String())
+		return nil, fmt.Errorf("method not found: %v", m.String())
+	}
+
+	// Verify required args
+	for k := range d.Required {
+		name := d.Required[k].Name
+		typ := d.Required[k].Type
+		arg := data.Get(name)
+		if arg.Type() != typ {
+			return nil, fmt.Errorf("%v: invalid type %v, want %v",
+				name, arg.Type(), typ)
+		}
+	}
+
+	return d, nil
+}
+
+func wasmVersion(_ js.Value, _ []js.Value) (js.Value, error) {
+	return Object{
+		"version":   version,
+		"gitCommit": gitCommit,
+	}.Value(), nil
+}
+
+func generateKey(_ js.Value, args []js.Value) (js.Value, error) {
+	log.Tracef("generatekey")
+	defer log.Tracef("generatekey exit")
+
+	var (
+		btcChainParams *btcchaincfg.Params
+		netNormalized  string
+	)
+	net := args[0].Get("network").String()
+	switch net {
+	case "devnet", "testnet3", "testnet":
+		btcChainParams = &btcchaincfg.TestNet3Params
+		netNormalized = "testnet3"
+	case "mainnet":
+		btcChainParams = &btcchaincfg.MainNetParams
+		netNormalized = "mainnet"
+	default:
+		return js.Null(), fmt.Errorf("invalid network: %v", net)
+	}
+
+	privKey, err := dcrsecpk256k1.GeneratePrivateKey()
+	if err != nil {
+		log.Errorf("failed to generate private key: %v", err)
+		return js.Null(), fmt.Errorf("generate secp256k1 private key: %w", err)
+	}
+	btcAddress, err := btcutil.NewAddressPubKey(
+		privKey.PubKey().SerializeCompressed(),
+		btcChainParams,
+	)
+	if err != nil {
+		log.Errorf("failed to generate btc address: %v", err)
+		return js.Null(), fmt.Errorf("create BTC address from public key: %v", err)
+	}
+
+	compressedPubKey := privKey.PubKey().SerializeCompressed()
+	ethereumAddress := ethereum.PublicKeyToAddress(compressedPubKey).String()
+
+	return Object{
+		"ethereumAddress": ethereumAddress,
+		"network":         netNormalized,
+		"privateKey":      hex.EncodeToString(privKey.Serialize()),
+		"publicKey":       hex.EncodeToString(compressedPubKey),
+		"publicKeyHash":   btcAddress.AddressPubKeyHash().String(),
+	}.Value(), nil
+}
+
+func startPoPMiner(_ js.Value, args []js.Value) (js.Value, error) {
+	log.Tracef("startPoPMiner")
+	defer log.Tracef("startPoPMiner exit")
+
+	pmMtx.Lock()
+	defer pmMtx.Unlock()
+	if pm != nil {
+		return js.Null(), errors.New("pop miner already started")
+	}
+
+	pm = new(Miner)
+	pm.ctx, pm.cancel = context.WithCancel(context.Background())
```

Comment:
Is this too late to create the context?

Reply:
This is copied from the previous code, where would be a better place for it?
I think this is the earliest that it can be stored in the global miner.

We could also initialise the Miner object at init and create the context in the main function? 🤔